// ng2-tree/src/tree.component.ts
import { Input, Component, OnInit, EventEmitter, Output, Inject, OnChanges, SimpleChanges } from '@angular/core';
import { TreeService } from './tree.service';
import { TreeModel, Ng2TreeSettings } from './tree.types';
import { NodeEvent } from './tree.events';
import { Tree } from './tree';
import { FoldingType } from './tree.types';

@Component({
  selector: 'app-tree',
  template: `<app-tree-internal [tree]="tree" [settings]="settings"></app-tree-internal>`,
  providers: [TreeService]
})
export class TreeComponent implements OnInit, OnChanges {
  private static EMPTY_TREE: Tree = new Tree({ value: '', fullPath: '', foldingType: FoldingType.Collapsed, hasChildren: false });

  @Input()
  public treeModel: TreeModel;

  @Input()
  public settings: Ng2TreeSettings;

  @Output()
  public nodeSelected: EventEmitter<any> = new EventEmitter();

  public tree: Tree;

  public constructor( @Inject(TreeService) private treeService: TreeService) {
  }

  public ngOnChanges(changes: SimpleChanges): void {
    if (!this.treeModel) {
      this.tree = TreeComponent.EMPTY_TREE;
    } else {
      this.tree = new Tree(this.treeModel);
    }
  }

  public ngOnInit(): void {
    this.treeService.nodeSelected$.subscribe((e: NodeEvent) => {
      this.nodeSelected.emit(e);
    });
  }
}
